syncer/db.go

```diff
@@ -69,45 +69,94 @@ type binlogSize struct {
+	if baseDB != nil {
+		err := baseDB.Close()
+		if err != nil {
+			tctx.L().Error("fail to close baseDB", log.ShortError(err))
+		}
 	}
-	return conn.baseConn.ResetConn(tctx)
 }
 
-func (conn *Conn) getMasterStatus(flavor string) (mysql.Position, gtid.Set, error) {
-	return utils.GetMasterStatus(conn.baseConn.DB, flavor)
+// UpStreamConn connect to upstream DB
+// Normally, we need to get some upstream information through some helper functions
+// these helper functions are all easy query functions, so we use a pool of connections here
+// maybe change to one connection some day
+type UpStreamConn struct {
+	BaseDB *conn.BaseDB
 }
 
-func (conn *Conn) getServerUUID(flavor string) (string, error) {
-	return utils.GetServerUUID(conn.baseConn.DB, flavor)
+func (conn *UpStreamConn) getMasterStatus(flavor string) (mysql.Position, gtid.Set, error) {
+	return utils.GetMasterStatus(conn.BaseDB.DB, flavor)
 }
 
-func (conn *Conn) getParser(ansiQuotesMode bool) (*parser.Parser, error) {
-	return utils.GetParser(conn.baseConn.DB, ansiQuotesMode)
+func (conn *UpStreamConn) getServerUUID(flavor string) (string, error) {
+	return utils.GetServerUUID(conn.BaseDB.DB, flavor)
 }
 
-func (conn *Conn) killConn(connID uint32) error {
-	return utils.KillConn(conn.baseConn.DB, connID)
+func (conn *UpStreamConn) getParser(ansiQuotesMode bool) (*parser.Parser, error) {
+	return utils.GetParser(conn.BaseDB.DB, ansiQuotesMode)
 }
 
-func (conn *Conn) fetchAllDoTables(bw *filter.Filter) (map[string][]string, error) {
-	return utils.FetchAllDoTables(conn.baseConn.DB, bw)
+func (conn *UpStreamConn) killConn(connID uint32) error {
+	return utils.KillConn(conn.BaseDB.DB, connID)
 }
 
-func (conn *Conn) countBinaryLogsSize(pos mysql.Position) (int64, error) {
-	return countBinaryLogsSize(pos, conn.baseConn.DB)
+func (conn *UpStreamConn) fetchAllDoTables(bw *filter.Filter) (map[string][]string, error) {
+	return utils.FetchAllDoTables(conn.BaseDB.DB, bw)
+}
+
+func (conn *UpStreamConn) countBinaryLogsSize(pos mysql.Position) (int64, error) {
+	return countBinaryLogsSize(pos, conn.BaseDB.DB)
+}
+
+func createUpStreamConn(dbCfg config.DBConfig) (*UpStreamConn, error) {
+	baseDB, err := createBaseDB(dbCfg)
+	if err != nil {
+		return nil, terror.WithScope(terror.DBErrorAdapt(err, terror.ErrDBDriverError), terror.ScopeUpstream)
+	}
+	return &UpStreamConn{BaseDB: baseDB}, nil
+}
+
+func closeUpstreamConn(tctx *tcontext.Context, conn *UpStreamConn) {
+	if conn != nil {
+		closeBaseDB(tctx, conn.BaseDB)
+	}
+}
+
+// DBConn represents a live DB connection
+// it's not thread-safe
+type DBConn struct {
+	cfg      *config.SubTaskConfig
+	baseConn *conn.BaseConn
+
+	// for workerConn to reset baseConn
+	resetConnFn func(*tcontext.Context) (*conn.BaseConn, error)
+}
+
+// ResetConn reset one worker connection from specify *BaseDB
+func (conn *DBConn) resetConn(tctx *tcontext.Context) error {
+	if conn == nil {
+		return terror.ErrDBDriverError.Generate("database not valid")
```

Comment:
Is `connection is nil` meaningful? _not valid_ provides less information.

Reply:
It can be remove, [Done](https://github.com/pingcap/dm/pull/266/files#diff-c3cc6fccd12f8e1679a575d1aaf8fd14R148)